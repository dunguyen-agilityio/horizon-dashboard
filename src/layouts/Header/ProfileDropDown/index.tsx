'use client';

// Libs
import { useState } from 'react';

// Components
import {
  Avatar,
  Modal,
  ModalBody,
  ModalContent,
  ModalFooter,
  ModalHeader,
  Popover,
  PopoverContent,
  PopoverTrigger,
  useDisclosure,
} from '@nextui-org/react';
import Text from '@/components/Text';
import Button from '@/components/Button';

// Types
import { TEXT_SIZE } from '@/types/text';
import { TImage } from '@/models/Image';

// Actions
import { handleSignOut } from '@/actions/auth';

interface IProfileDropdownProps {
  avatar?: TImage;
  username: string;
}

const ProfileDropDown = ({ avatar, username }: IProfileDropdownProps) => {
  const [isVisible, setIsVisible] = useState(false);

  const { isOpen, onOpen, onClose } = useDisclosure();

  const handleClose = () => {
    onClose();
  };

  const handleCloseDropDown = () => {
    setIsVisible((prev) => !prev);
  };

  const handleLogout = async () => {
    await handleSignOut();
    handleClose();
  };

  const handleCloseModal = () => {
    onOpen();
    handleCloseDropDown();
  };

  return (
    <div>
      <Popover
        placement="bottom-end"
        showArrow={true}
        isOpen={isVisible}
        onOpenChange={handleCloseDropDown}
        className="before:bg-gray before:dark:bg-indigo-light"
      >
        <PopoverTrigger>
          <button
            className="focus-visible:outline-none aria-expanded:scale-1 aria-expanded:opacity-100"
            data-testid="profile-trigger-btn"
          >
            <Avatar src={avatar?.url} alt={username} />
          </button>
        </PopoverTrigger>
        <PopoverContent
          className="bg-gray dark:bg-indigo-light"
          data-testid="profile-content"
        >
          <div className="flex flex-col gap-4 px-1 py-2 min-w-[150px]">
            <Text size={TEXT_SIZE.md} className="font-bold">
              {username}
            </Text>
            <Button
              data-testId="button-sign-out"
              className="justify-start"
              variant="flat"
              onClick={handleCloseModal}
            >
              Sign Out
            </Button>
          </div>
        </PopoverContent>
      </Popover>
      {isOpen && (
        <Modal
          isOpen
          onOpenChange={onClose}
          isDismissable={false}
          data-testid="confirm-signOut-modal"
        >
          <ModalContent className="bg-white dark:bg-content1">
            <ModalHeader className="flex flex-col gap-1">Sign Out</ModalHeader>
            <ModalBody>
              <Text>Are you sure you want to sign out?</Text>
            </ModalBody>
            <ModalFooter>
              <Button color="danger" variant="light" onPress={handleClose}>
                Close
              </Button>
              <Button
                data-testId="modal-button-sign-out"
                color="primary"
                onPress={handleLogout}
              >
                Sign Out
              </Button>
            </ModalFooter>
          </ModalContent>
        </Modal>
      )}
    </div>
  );
};

export default ProfileDropDown;
