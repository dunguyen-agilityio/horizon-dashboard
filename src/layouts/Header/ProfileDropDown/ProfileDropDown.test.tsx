import {
  act,
  fireEvent,
  render,
  queryByTestId,
  getByText,
  waitFor,
} from '@testing-library/react';
import ProfileDropDown from '.';

describe('ProfileDropDown tests', () => {
  it('Should match snapshot', () => {
    const { container } = render(<ProfileDropDown />);
    expect(container).toMatchSnapshot();
  });

  it('Should show profile DropdownContent when click Trigger button', () => {
    const { getByTestId } = render(<ProfileDropDown />);
    expect(
      queryByTestId(document.documentElement, 'profile-content'),
    ).not.toBeInTheDocument();

    act(() => {
      fireEvent.click(getByTestId('profile-trigger-btn'));
    });

    expect(
      queryByTestId(document.documentElement, 'profile-content'),
    ).toBeInTheDocument();
  });

  it('Should show ConfirmLogoutModal when click Logout button', () => {
    const { getByTestId } = render(<ProfileDropDown />);

    act(() => {
      fireEvent.click(getByTestId('profile-trigger-btn'));
    });

    act(() => {
      fireEvent.click(getByText(document.documentElement, 'Logout'));
    });

    expect(
      queryByTestId(document.documentElement, 'confirm-logout-modal'),
    ).toBeInTheDocument();
  });

  it('Should show Hidden ConfirmLogoutModal when click Close button', async () => {
    const { getByTestId } = render(<ProfileDropDown />);

    act(() => {
      fireEvent.click(getByTestId('profile-trigger-btn'));
    });

    act(() => {
      fireEvent.click(getByText(document.documentElement, 'Logout'));
    });

    fireEvent.click(getByText(document.documentElement, 'Close'));

    await waitFor(() => {
      expect(
        queryByTestId(document.documentElement, 'confirm-logout-modal'),
      ).not.toBeInTheDocument();
    });
  });

  it('Should show Hidden ConfirmLogoutModal when click Logout button', async () => {
    const { getByTestId } = render(<ProfileDropDown />);

    act(() => {
      fireEvent.click(getByTestId('profile-trigger-btn'));
    });

    act(() => {
      fireEvent.click(getByText(document.documentElement, 'Logout'));
    });

    const modal = queryByTestId(
      document.documentElement,
      'confirm-logout-modal',
    );

    fireEvent.click(getByText(modal!, 'Logout'));

    await waitFor(() => {
      expect(modal).not.toBeInTheDocument();
    });
  });
});
