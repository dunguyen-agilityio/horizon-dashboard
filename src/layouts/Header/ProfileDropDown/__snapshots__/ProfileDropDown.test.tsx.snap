// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileDropDown tests Should match snapshot 1`] = `
<div>
  <button
    aria-expanded="false"
    class="z-10 subpixel-antialiased focus-visible:outline-none aria-expanded:scale-1 aria-expanded:opacity-100"
    data-slot="trigger"
    data-testid="profile-trigger-btn"
    type="button"
  >
    <img
      alt="Adela Parkson"
      class="min-w-10 rounded-full"
      data-nimg="1"
      decoding="async"
      height="40"
      loading="lazy"
      src="/_next/image?url=https%3A%2F%2Fres.cloudinary.com%2Fdn5vw6fjp%2Fimage%2Fupload%2Fv1726630608%2Fimage_de1e29b0b7.png&w=96&q=75"
      srcset="/_next/image?url=https%3A%2F%2Fres.cloudinary.com%2Fdn5vw6fjp%2Fimage%2Fupload%2Fv1726630608%2Fimage_de1e29b0b7.png&w=48&q=75 1x, /_next/image?url=https%3A%2F%2Fres.cloudinary.com%2Fdn5vw6fjp%2Fimage%2Fupload%2Fv1726630608%2Fimage_de1e29b0b7.png&w=96&q=75 2x"
      style="color: transparent;"
      width="40"
    />
  </button>
</div>
`;
